Allow Template upgrade properties to be set and remove properties that no longer exist.

JC-wk opened the pull request:
# Resolves #4732 #4730

## What is being addressed

Currently if you add a new property to a template there is no way to specify this property before an upgrade is ran.
The upgrade may fail due to the missing property (although the template version is still incremented)
The user then has to click update and supply the property.
Similarly when removing a property from a template and running an upgrade, the property still exists on the resource.

# Todo
- [x] Enhancement: Current template should only be retrieved once.
- [x] To check #4795 and fix issue #4728
- [x] To allow us to merge in the Unit Test #4794 first so the UI tests can be added
- [x] Possibly handle cases such as where an enum value is removed (not the field entirely) and that value is currently chosen

## How is this addressed

- Adds a form to allow the user to specify new properties prior to the upgrade
- Removes any template properties that no longer exist in the new template
- Added tests
- Updated CHANGELOG.md 
- Increment API version

<img width="347" height="611" alt="image" src="https://github.com/user-attachments/assets/a0971b41-2573-4a43-9fe4-084bb6f18e06" />
<img width="425" height="374" alt="image" src="https://github.com/user-attachments/assets/8a8a434a-7137-40ba-88bb-103630657037" />

Commit:
[WIP] Submitting new properties during upgrade

google-labs-jules[bot] committed:
fix(ui): Handle conditional properties in upgrade form

This commit aligns the resource upgrade process with the update process by correctly handling conditional properties in the JSON schema.

- The schema generation logic in `ConfirmUpgradeResource.tsx` is updated to include conditional blocks (`if`/`then`/`else`) when the condition is based on an existing property.
- New read-only properties are now submitted during the upgrade process.

google-labs-jules[bot] committed:
fix(ui): Handle conditional properties in upgrade form

This commit aligns the resource upgrade process with the update process by correctly handling conditional properties in the JSON schema.

- The schema generation logic in `ConfirmUpgradeResource.tsx` is updated to include conditional blocks (`if`/`then`/`else`) when the condition is based on an existing property.
- New read-only properties are now submitted during the upgrade process.
- The `liveOmit` prop is added to the form to prevent the submission of unevaluated properties from conditionally hidden fields.

JC-wk committed:
Merge pull request #1 from JC-wk/fix/upgrade-conditional-properties-11346005040390942732

Fix Upgrade Conditional Properties

Commit:
support all types of template

Commit:
show properties to be removed

Commit:
remove logging

Commit:
update warning message

Commit:
add capability to remove properties that no longer exist in the template

Commit:
add tests and resolve issue causing failing test

Commit:
update changelog

Commit:
cache currrent template

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
Add test

Commit:
fix ws template auth

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
ui version bump

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
bump version

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
fix: use correct auth
fix: allow upgrades on hidden properties

Commit:
test: enhance ConfirmUpgradeResource tests for template fetching and property visibility

Commit:
update test

Commit:
add method to extract pipeline properties for resource validation

Commit:
add tests

Commit:
update api version

JC-wk committed:
Add ESLint config file paths for validation

JC-wk committed:
Merge branch 'microsoft:main' into main

Commit:
allow property substitution without an update to the parent workspace in the pipeline

Commit:
use .allOf property

JC-wk committed:
Merge branch 'microsoft:main' into main

Commit:
Merge remote-tracking branch 'TREfork/main' into template-upgrade-properties

Commit:
Merge remote-tracking branch 'TREfork/main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
update versions

Commit:
filter out properties set via pipeliine

Commit:
revert allOf change

Commit:
revert tests

Commit:
correctly revert tests

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
changelog

Commit:
api ver

Commit:
fix lint

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
update version

JC-wk commented on ui/app/src/components/shared/ConfirmUpgradeResource.tsx:
```diff
@@ -23,24 +26,114 @@ import { LoadingState } from "../../models/loadingState";
+    }
+  }
+  return keys;
+};
```
If we had used the simplified  Object.keys(properties)  approach (Option 2) suggested by Copilot:               
                                                                                                                  
  • An upgrade from a template with  parent: { child }  to  parent: { child, sibling }  would result in both      
  versions having the top-level key  parent .                                                                     
  • Comparing them would yield  []  (no new properties detected).                                                 
  • The upgrade would proceed silently without giving the user a chance to supply a value for the new  sibling    
  property, causing backend validation failures or incomplete deployments.                                        
                                                                                                                  
  By keeping the recursive dotted-path detection but updating the form state management, validation, and API call 
  generation to handle nested objects correctly:
  
  1. Accurate Detection: We correctly identify when a nested property (e.g.,  parent.sibling ) is added/removed.  
  2. Object Structure Alignment: We keep  newPropertyValues  in the nested object shape that the RJSF  Form  and  
  the backend PATCH API expect (e.g.,  { parent: { child: '...', sibling: '...' } } ).
  3. Data Preservation: We pre-populate existing sibling values (e.g., keeping  parent.child ) so they aren't lost
  or cleared during the upgrade.
  4. Validation Resolution: We use ConfirmUpgradeResource.tsx to check if the new nested fields are filled,
preventing the    
  "Upgrade" button from being permanently disabled.

Commit:
Fix PR suggestions from copilot

Enhance template upgrade functionality and improve resource handling

- Allow new template properties to be specified during

upgrades and remove obsolete properties.
- Update resource repository methods to handle parent service names for user resources.
- Add tests for nested properties and enum value handling in upgrade scenarios.
- Refactor ConfirmUpgradeResource component to manage nested properties and validation more effectively.
- Update TypeScript configuration for improved compatibility.

Commit:
apply copilot sec fix

JC-wk commented on ui/app/src/components/shared/ConfirmUpgradeResource.tsx:
```diff
@@ -23,24 +26,207 @@
+    }
+    current = current[part];
+  }
+  current[parts[parts.length - 1]] = value;
```
fix applied

Commit:
implplement copilot suggestions

Commit:
Guarded the  allOf  attribute lookup in resources.py using Python's  hasattr  built-in function:
        if hasattr(resource_template, "allOf") and resource_template.allOf is not None:

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Bump version to 0.25.27

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
Bump version to 0.8.28 in package.json and package-lock.json

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

marrobi commented:
# Review: PR #4783 — Template upgrade properties

Solid, well-tested feature. The approach (fetch target schema, prompt for new properties, strip removed ones) is sound. Below are the issues grouped by severity, with notes on where they overlap with existing bot comments.

## 🔴 High – correctness bugs

### 1. `_get_all_property_keys_from_template` — `allOf` detection on a Pydantic model (`resources.py`)
```python
if hasattr(resource_template, "allOf") and resource_template.allOf is not None:
```
`ResourceTemplate` is a Pydantic model. Unless `allOf` is a declared field, `hasattr` is unreliable and `resource_template.allOf` won't reflect the raw conditional schema. Meanwhile the **frontend** and `validate_patch` operate on the enriched **dict**. This method operates on the model — an inconsistency worth resolving. Normalize to a dict (e.g. `.dict()` / the enriched template) before extracting keys, so both the removal path and validation path see conditional properties identically. Aligns with bot comments r3379672301 and r3379888936.

### 2. Removed-property deletion only handles top-level keys, but the UI warns about nested paths
Backend `patch_resource`:
```python
for prop in properties_to_remove:
    if prop in resource.properties:
        del resource.properties[prop]
```
`properties_to_remove` yields top-level keys, while the **UI** computes `removedProperties` from the recursive dotted-path list and shows them to the user. The two can diverge: the UI may warn `parent.child` will be removed, but the backend only deletes top-level keys. Either:
- make the UI warning use top-level keys only (matching backend), **or**
- make the backend delete nested paths too.

Pick one and keep them symmetric. Agrees with bot comment r3379888875. (The author's rationale in r3379774425 defends recursive *detection of new* properties — fine — but the *removal* warning symmetry is a separate, still-open concern.)

### 3. UserResource upgrades need `parent_service_name` — verify the source works, not just the tests
Both new `get_template_by_name_and_version` calls now pass:
```python
parent_service_name = getattr(resource_template, "parentWorkspaceService", None)
```
Tests assert this is forwarded, addressing bot comments r3379672323 / r3379672353. But tests set `parentWorkspaceService` to a string. Confirm that in production `resource_template.parentWorkspaceService` is actually populated on the **old** template model at this point. If it can be `None` for a real UserResource, the lookup still raises. Worth an integration check or a guard.

## 🟠 Medium – behavior / UX

### 4. Optional new properties block the Upgrade button
The disable predicate now checks `isPropertyRequiredInState`, an improvement over the earlier "any empty visible property" logic (bot comment r3379672266). Good. Double-check the enum-invalid branch doesn't inadvertently block on *optional* fields left at their valid existing value — the `val !== undefined && val !== ""` guard looks right; keep the enum test to lock it in.

### 5. Prefill uses schema defaults only — good, but confirm `''` isn't injected elsewhere
Final code sets values only when `propSchema.default !== undefined` (addresses bot comment r3379672242). With `omitExtraData` / `liveOmit`, optional untouched fields should be omitted from the PATCH body. Confirm boolean/number/array-typed new properties round-trip without being coerced to `""`. A test with a non-string new property would harden this.

## 🟢 Low – nits / housekeeping

- **CHANGELOG** (bot r3379672384): entry now includes `([#4783](...))` — appears resolved ✅
- **Prototype pollution** (CodeQL r3379843396): `partGuard` guards `__proto__`/`constructor`/`prototype`; author replied "fix applied." Confirm alert #56 is resolved on the latest commit ✅
- **Flaky test** (bot r3379672411): "displays error" test was fixed to click once inside `act(...)` then `waitFor` the assertion. Scan the other upgrade tests to ensure none still `fireEvent.click` inside a `waitFor` callback.
- **`any` creep**: component and tests introduce many `any` casts. Consider typing `newPropertyValues` as `Record<string, unknown>` at minimum.
- **`tsconfig.json`**: `"ignoreDeprecations": "6.0"` added — add a comment noting which deprecated option it silences so a future TS bump can remove it. The array reformatting is harmless noise.
- **Dead/commented code** in `ConfirmUpgradeResource.tsx`: remove `// let workspaceApplicationIdURI = undefined;` and related commented lines before merge.
- **Duplicated key-extraction logic**: `getAllPropertyKeysFromTemplate` (TS) and `_get_all_property_keys_from_template` (Py) reimplement the same traversal. Extract the TS helpers out of the `useEffect` for testability and readability.

## Summary

Core mechanism is good and test coverage is strong. Before merge, prioritize:
1. Reconciling the `allOf` / model-vs-dict inconsistency in `_get_all_property_keys_from_template` (High #1).
2. Making the UI "removed properties" warning symmetric with what the backend actually deletes (High #2).
3. Confirming real UserResource upgrades supply `parent_service_name` in production, not just in mocks (High #3).

Then the low-severity cleanups. Most Copilot-flagged items (CHANGELOG ref, prototype pollution, required-only disable logic, prefill defaults) appear already addressed in the latest revision.

Commit:
formatting

Commit:
Merge remote-tracking branch 'TREfork/main' into template-upgrade-properties

Commit:
address pr feedback

Commit:
bump version to 0.25.30

Commit:
Updated test_resource_repository.py to align template os_image options with current template definitions:

  • Template Fixtures Updated: Replaced obsolete OS names (Windows 10, Server 2019 Data Science VM, Server 2022 Data Science VM) with active OS options (Windows 11, Windows
  Server 2025) in sample_resource_template() and sample_resource_template_with_version().
  • Upgrade Test Updated: Updated test_validate_patch_rejects_existing_non_updateable_property_during_upgrade to use Windows Server 2025 for os_image.

Commit:
1. Comprehensive old_template_properties Extraction:
      • old_template_properties is now computed using self._get_all_property_keys_from_template(resource_template), which inspects both root properties and allOf (then.
      properties / else.properties) on the old template object or dictionary.
  2. Explicit Per-Property Permission Validation:
      • Implemented get_prop_schema inside validate_patch to resolve property definitions across both root properties and allOf conditional blocks.
      • Every property in resource_patch.properties is explicitly validated against allowed rules:
        is_updateable = prop_def.get("updateable", False) is True if prop_def else False
        is_new_on_upgrade = is_upgrade and prop_name not in old_template_properties
        is_pipeline_prop = prop_name in pipeline_properties

        if not (is_updateable or is_new_on_upgrade or is_pipeline_prop):
            raise ValidationError(f"Property '{prop_name}' is not updateable.")

  3. Added Unit Test:
      • Added test_validate_patch_rejects_non_updateable_allOf_property in test_resource_repository.py to verify that non-updateable conditional properties in allOf are
      correctly rejected during patch calls.

Commit:
Enhance resource template property handling during upgrades

- Updated `_get_all_property_keys_from_template` to support nested properties with prefixes.
- Modified `get_prop_schema` to retrieve property definitions using dotted paths.
- Introduced `is_property_allowed` to validate properties during patching, allowing new nested properties under existing non-updateable objects.
- Added unit test to verify behavior for new nested properties during template upgrades.

Commit:
Refactor schema upgrade utilities and enhance ConfirmUpgradeResource component for better property handling

Commit:
Implement nested property removal in patch_resource and add E2E tests for template upgrades

Commit:
Enhance resource patch validation to support nested properties and improve loading state management in ConfirmUpgradeResource component

Commit:
Update test to validate patch with Windows 11 OS image for resource properties

Commit:
updated the test to use an admin token when fetching the workspace service template to avoid unauthorized GETs against admin-only template endpoints.

Commit:
fix copilot suggestions

Commit:
address copilot comments

Commit:
Refactor tests to remove unnecessary mocks for resource template repository and improve clarity in patch validation tests. Update ConfirmUpgradeResource component to streamline error handling and remove redundant workspace ID assignment.

Commit:
Refactor collectConditionalKeys to use type guard for required property filtering

Commit:
rem e2e

Commit:
copilot suggestion

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

JC-wk committed:
Merge branch 'main' into template-upgrade-properties

Commit:
### Summary of Changes

  #### 1. Early Error Return Loading States in UI

  • Files: ConfirmUpgradeResource.tsx
  • Fix: Added setRequestLoadingState(LoadingState.Error) before early returns when:
      1. Workspace context is missing for a UserResource (line 147)
      2. Parent workspace service info is missing (line 157)
      3. Resource type is unsupported (line 165)
  This ensures that ExceptionLayout.tsx renders properly instead of leaving the dialog stuck in LoadingState.Ok.

  #### 2. Schema Pruning & Nested Property Prefilling

  • Files: schemaUpgradeUtils.ts, ConfirmUpgradeResource.tsx, and ConfirmUpgradeResource.test.tsx
  • Fix:
      • Implemented pruneSchemaNode in schemaUpgradeUtils.ts to recursively prune object schemas so buildReducedSchema only includes
      newly-added or modified sub-fields instead of rendering existing, unmodifiable sub-properties.
      • Refactored initialValues prefilling in ConfirmUpgradeResource.tsx to populate values only for target missing/updated property
      keys rather than deep-cloning whole top-level parent objects (preventing removed or read-only properties from being re-sent).
      • Updated test assertions in ConfirmUpgradeResource.test.tsx and added new test cases for early error return paths.

  #### 3. Duplicate Template Fetches in Backend

  • Files: resources.py
  • Fix:
      • Updated validate_template_version_patch to return the fetched ResourceTemplate.
      • Reused new_template in patch_resource and passed target_template=new_template into validate_patch, eliminating duplicate
      database reads for the template during resource upgrade PATCH operations.

  #### 4. Install Pipeline Test Fixture

  • Files: test_resource_repository.py
  • Fix: Updated test_validate_patch_allows_install_pipeline_property to use 'install' instead of 'upgrade' under
  template_dict['pipeline'].
  ──────
  ### Verification

  • UI Tests: Ran npx vitest run in ui/app — 27/27 test files (259 tests) passed.
  • Backend Tests: Ran pytest in api_app — 737/737 tests passed.

Commit:
address copilot review

Commit:
1. Enum Property Value Prefilling (ConfirmUpgradeResource.tsx):
      • initialValues pre-filling now checks whether an existing property value is allowed by the target template's enum schema. If
      invalid for the new enum, it prefers the template default (if defined) or leaves the value unset rather than preserving an invalid
      enum option.
  2. Enum-Invalid Keys Visibility & User Input (ConfirmUpgradeResource.tsx):
      • visibleNewPropKeys now keeps keys with invalid enum values in the list of properties needing user input regardless of tre-hidden.
      • sanitizedUiSchema strips tre-hidden from uiSchema for any property requiring user input so the input control is rendered visibly
      in the form for user interaction.
  3. Prototype Pollution Protection (schemaUpgradeUtils.ts):
      • Added partGuard checks (__proto__, constructor, prototype) to getAllPropertyKeys and getTopLevelKeysFromTemplate to ensure
      consistent prototype pollution safety across schema utilities.

Commit:
feat: implement deep dictionary update method and add corresponding tests

Commit:
1. ui/app/src/components/shared/ConfirmUpgradeResource.tsx & ui/app/src/utils/schemaUpgradeUtils.ts
      • Issue: schemaUpgradeUtils.ts was evaluating conditional allOf schema rules using only newPropertyValues. If a conditional
      requirement depended on existing resource properties already on the resource, the condition would evaluate to false and allow
      upgrading without requiring the newly required field.
      • Fix: Added schemaUpgradeUtils.ts to deeply merge props.resource.properties with newPropertyValues. Updated the Upgrade button's
      ConfirmUpgradeResource.tsx check in ConfirmUpgradeResource.tsx to pass this combined state into isPropertyRequiredInState.
      • Test: Added unit test in ConfirmUpgradeResource.test.tsx to verify that conditional required rules based on existing resource
      properties properly disable the Upgrade button until filled.
  2. api_app/db/repositories/resources.py
      • Issue: resources.py returned the original resource_template object even when templateVersion was patched and new_template was
      fetched and validated, causing downstream callers to receive a stale template object.
      • Fix: Updated line 212 of resources.py to return new_template if new_template is not None else resource_template.
      • Test: Updated test_resource_repository.py to assert that patch_resource returns new_template on upgrade.

Commit:
add test

Commit:
fix: handle component unmount during schema fetch to prevent state updates

Commit:
fix: enhance validation for resource patch properties to distinguish between unexpected and non-updateable fields

Commit:
fix: update modal blocking behavior and enhance effect dependencies for resource upgrade

Commit:
fix: enhance logic for identifying removed properties in resource upgrade

Commit:
fix: update get_nested_val return type and enhance validation logic for resource patch properties

Commit:
fix: implement deep merge for resource patch properties and enhance validation logic

Commit:
fix: enhance property merging logic to include system properties and validate current properties against target template

Commit:
fix: enhance validation logic to reject modifications of system properties during template upgrades

Commit:
test: add validation for enum property updates during template upgrades

Commit:
fix: update _get_pipeline_properties to use set for unique property names and enhance validation in tests

Commit:
fix: enhance resource patch validation to allow retained system properties during upgrades

Commit:
test: add validation to reject empty object for non-updateable property in resource patch

Commit:
feat: enhance resource upgrade logic and validation for new properties in templates

Commit:
feat: add form error handling and validation for resource upgrade properties

Commit:
fix: update nested value setting logic to handle required properties in upgrade form

Commit:
feat: add additional fields to resource template schema for enhanced metadata

Commit:
feat: enhance property validation logic in ConfirmUpgradeResource component
